Add `@storybook/nextjs` framework

kylegach opened the pull request:
## What I did

> **Note**: Because our current workflow requires this PR to be merged before we can consistently check it in CI, this work is "feature flagged" to avoid inadvertent use by users of `sb@next init`. See: https://github.com/storybookjs/storybook/pull/19478.

- Add `@storybook/nextjs` framework
- Add nextjs/default-js and nextjs/default-ts to repro configs
- Add basic NEXTJS project type to CLI
    - Includes framework-specific template files
- Update CLI logic
    - Add framework option to baseGenerator
    - Add `hasFrameworkTemplates`
        - Ensures frameworks with templates use those instead of templates for associated renderer
        - Follows pattern of `hasInteractiveStories`
    - Update `copyComponents` to copy common template files before (instead
  of after) framework/render template files
        - Allows framework/renderer template files to contain, e.g. `Introduction.stories.mdx` and have it not be overwritten
- Apply STORYBOOK_REPRO_GENERATOR to env when running `sb init` in repro generator
    - Serves as a "feature flag" around detection of the framework in CLI

> **Note**: This work is _heavily_ based on [storybook-addon-next](https://github.com/RyanClementsHax/storybook-addon-next) by @RyanClementsHax and [storybook-addon-next-router](https://github.com/lifeiscontent/storybook-addon-next-router) by @lifeiscontent. Thank you for your efforts!

## How to test

1. `yarn generate-repros-next --template nextjs/default-js --local-registry`
1. `yarn task --task sandbox --template nextjs/default-js --from-local-repro`
1. Repeat with `nextjs/default-ts`

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [x] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

kylegach committed:
Add `nextjs/default-js` and `nextjs/default-ts` to repro configs

kylegach committed:
Add basic NEXTJS project type to CLI

kylegach committed:
Add `nextjs` framework

Co-authored-by: Yann Braga <yannbf@gmail.com>

yannbf committed:
add framework option to baseGenerator

kylegach committed:
Update `baseGenerator` to use framework's template

- Remove `frameworkToRenderer`
    - This information is already available from `baseGenerator`'s arguments
- Add `hasFrameworkTemplates`
    - Ensures frameworks with templates use those instead of templates for associated renderer
    - Follows pattern of `hasInteractiveStories`
- Update `copyComponents` to copy common template files before (instead
  of after) framework/render template files
    - Allows framework/renderer template files to contain, e.g. `Introduction.stories.mdx` and have it not be overwritten

kylegach committed:
Add nextjs framework template files

kylegach commented on code/frameworks/nextjs/README.md:
```diff
@@ -0,0 +1,622 @@
+# Storybook for Next.js <!-- omit in toc -->
```
This README's content was mostly copied (with minor adaptations) from https://github.com/RyanClementsHax/storybook-addon-next/blob/main/README.md

kylegach commented on code/frameworks/nextjs/README.md:
```diff
@@ -0,0 +1,622 @@
+
+### next.config.js
+
+#### ESM
```
Is this still true now that this is a framework?

yannbf replied:
Hey @RyanClementsHax ! I added support for esm in [this commit](https://github.com/storybookjs/storybook/pull/19382/commits/252fb59205b3515a30076539891fb46892bd5b00), in case you're interested!

RyanClementsHax replied:
🤯 I can't believe I forgot about 'import'!

kylegach commented on code/frameworks/nextjs/package.json:
```diff
@@ -0,0 +1,129 @@
+    ],
+    "platform": "node"
+  }
+}
```
Other frameworks have a `gitHead` property. Is that auto-generated by some release process?

kylegach commented on code/lib/cli/rendererAssets/nextjs/js/Button.jsx:
```diff
@@ -0,0 +1,54 @@
+      {...props}
+    >
+      {label}
+      <style jsx>{`
```
This is exactly like the React version of this template file, with the inline style swapped for styled-jsx, Next.js' preferred (CSS-in-JS) styling solution.

tmeasday replied:
@kylegach is it guaranteed that a user will have `styled-jsx` installed if they are using `next`? I don't think we do in our monorepo... 🤔

kylegach replied:
It worked as expected when running the Storybook generated by the `yarn generate-repros-next --template nextjs/default-js --local-registry` command, so I think the answer is "yes".

yannbf replied:
`styled-jsx` comes out of the box with Nextjs:
https://nextjs.org/blog/styling-next-with-styled-jsx

kylegach commented on code/lib/cli/rendererAssets/nextjs/js/Introduction.stories.mdx:
```diff
@@ -0,0 +1,228 @@
+      margin-bottom: 2px;
+    }
+
+    .link-item-img-wrapper {
```
This is exactly like the common version, but with this selector updated (was: `.link-item img`) and making use of `next/image`. Because of how the framework handles image imports, the images were broken in the original version. And the use of `next/image` broke the styling, hence the selector update.

kylegach commented on code/lib/cli/src/generators/baseGenerator.ts:
```diff
@@ -103,12 +97,16 @@ const stripVersions = (addons: string[]) => addons.map((addon) => getPackageDeta
 const hasInteractiveStories = (rendererId: SupportedRenderers) =>
   ['react', 'angular', 'preact', 'svelte', 'vue', 'vue3', 'html'].includes(rendererId);
 
+const hasFrameworkTemplates = (framework?: SupportedFrameworks) =>
```
We needed a way to identify the frameworks with framework-specific template files.

kylegach commented on code/lib/cli/src/helpers.ts:
```diff
@@ -212,10 +215,8 @@ export async function copyComponents(renderer: SupportedRenderers, language: Sup
   };
 
   const destinationPath = await targetPath();
+  await fse.copy(join(getBaseDir(), 'rendererAssets/common'), destinationPath, { overwrite: true });
```
Swapped the order here so that a framework/renderer could contain a file that is otherwise covered by the common template files, but the specific one would win.

RyanClementsHax commented on code/frameworks/nextjs/src/resolved-router-context.ts:
```diff
@@ -0,0 +1,12 @@
+// this file will be aliased by webpack at runtime (this is just for typing)
+
+import { RouterContext as RouterContextV12 } from 'next/dist/shared/lib/router-context';
+import { RouterContext as RouterContextV11 } from 'nextv11/dist/next-server/lib/router-context';
```
I'd recommend taking out the extra types. Including the past versions of nextjs fires all kinds of alarms for security scans even though you only use it for type checking. Integration tests, I've found, are perfectly fine to ensure proper usage of the router.

kylegach replied:
An attempt at this is here: https://github.com/storybookjs/storybook/commit/af0477dcc4602a9d38435d11568f5981cf8b3012

RyanClementsHax commented on code/lib/cli/src/detect.test.ts:
```diff
@@ -139,6 +139,16 @@ const MOCK_FRAMEWORK_FILES: {
+    files: {
+      'package.json': {
+        dependencies: {
+          next: '^9.0.0',
```
curious: why version 9?

kylegach replied:
Why 9 in general? Because your addon from which this is ~copy/pasted~ _based_ already supported it.

Why 9 right here? To test the version range detection.

RyanClementsHax commented on code/frameworks/nextjs/README.md:
```diff
@@ -0,0 +1,622 @@
+
+### Next.js's Image Component
+
+[next/image](https://nextjs.org/docs/api-reference/next/image) is [notoriously difficult](https://github.com/vercel/next.js/issues/18393) to get working with Storybook. This framework allows you to use Next.js's `Image` component with no configuration!
```
also heads up: before releasing this framework, consider including support for `next/future/image`. I've been getting a bunch of github issues to support it. The workaround I do in my addon for the regular image should be a cookie cutter solution for the future image. (There's also an open issue on my addon's repo with some other workarounds people made)

kylegach replied:
For reference: https://github.com/RyanClementsHax/storybook-addon-next/issues/99

kylegach replied:
I suspect the main difficulty here is going to be conditionally including whatever code ends up being used from the above reference (e.g. we can't `require('next/future/image')` if the project is using a version of `next` that doesn't include that package).

Additionally, it's not likely imo, but we may need to update the selector [mentioned here](https://github.com/storybookjs/storybook/pull/19382#discussion_r989321448) to work with both `next/image` (which includes wrapping markup around the `img`)  _and_ `next/future/image` (which does not).

RyanClementsHax commented:
question: what is the current position on supporting the vite builder? I've gotten a few issues on my addon's repo for vite support.

kylegach commented:
> question: what is the current position on supporting the vite builder? I've gotten a few issues on my addon's repo for vite support.

Isn't Next Webpack-only? I don't understand why people would ask for that. I'm away from my computer at the moment; will look into those issues later for my own edification.

Thanks for taking a look, Ryan!

RyanClementsHax commented:
> > question: what is the current position on supporting the vite builder? I've gotten a few issues on my addon's repo for vite support.
> 
> Isn't Next Webpack-only? I don't understand why people would ask for that. I'm away from my computer at the moment; will look into those issues later for my own edification.
> 
> Thanks for taking a look, Ryan!

Yup. It is. It seems there is a hunger to use the vite builder just for storybook though. I imagine this is possible too. I don’t know how much of the code can be shared however :/

kylegach commented on code/frameworks/nextjs/README.md:
```diff
@@ -0,0 +1,622 @@
+
+## Requirements
+
+- [Next.js](https://nextjs.org/) >= 9.x
```
Yes. Honestly, that's only because your addon already had support and it seemed straightforward to continue supporting.

I'm curious if other reviewers have opinions on the tradeoffs of broad support vs. less code to maintain, though...

RyanClementsHax replied:
I can say that supporting v9 is definitely doable. It is probably a good idea to keep v9 support at least in the first major version of this so that people can more easily swap between my addon and this addon.

On the same line of thinking it probably is a good idea to make a point of supporting all the same features just to reduce friction from swapping 🤔

shilman replied:
We use the Next Image in our CLI template files. Since Next v9 does not support Next Images, we shouldn't try to support it and complicate our setup. @kylegach @yannbf

himyjan commented:
`@storybook/react-vite` builder work well with next.js
[Storybook v7 example](https://github.com/himyjan/Storybook-v7-Alpha-Vite3-webpack5-Angular15next-React18-Vue3-TypeScript)

kylegach commented on code/frameworks/nextjs/src/utils.ts:
```diff
@@ -0,0 +1,55 @@
+ * // after main script path truncation
+ * scopedResolve('styled-jsx') === '/some/path/node_modules/styled-jsx'
+ */
+export const scopedResolve = (id: string): string => {
```
Possible alternative: https://github.com/storybookjs/storybook/blob/d35ed61320dca94e26387bccf01e3ee6a93a9e0c/code/lib/cli/src/dirs.ts#L1-L5

kylegach committed:
Remove extra types

See: https://github.com/storybookjs/storybook/pull/19382/#discussion_r989365320

RyanClementsHax commented on code/frameworks/nextjs/src/css/webpack.ts:
```diff
@@ -0,0 +1,108 @@
+import semver from 'semver';
+import { scopedResolve } from '../utils';
+
+export const configureCss = (baseConfig: WebpackConfig, nextConfig: NextConfig): void => {
```
just a heads up: when developing this, this file had a large amount of code churn because nextjs's css configuration is very specific. Originally, I thought a lot of the configuration was extraneous for the purposes of storybook, but many people depend on the various things that the nextjs css config supports. It is worth keeping the nextjs css config files bookmarked.

RyanClementsHax commented on code/frameworks/nextjs/src/css/webpack.ts:
```diff
@@ -0,0 +1,108 @@
+ * https://webpack-3.cdn.bcebos.com/loaders/css-loader/#url
+ * https://webpack-3.cdn.bcebos.com/loaders/css-loader/#import
+ */
+const getImportAndUrlCssLoaderOptions = (nextConfig: NextConfig) =>
```
its worth checking if with the upgrade of storybook to v7 we can get rid of css loader v4 support.

RyanClementsHax replied:
Maybe.... it wasn't that some people were using webpack 4. My addon never supported webpack 4. It was just that some people because of npm versiom resolution were using various css loader major versions. I believe part of the problem was from some storybook internal dependencies but I can't remember.

RyanClementsHax replied:
https://github.com/RyanClementsHax/storybook-addon-next/issues/22 here is one of the related issues

kylegach committed:
Feature flag CLI framework detection

- Apply STORYBOOK_REPRO_GENERATOR to env when running `sb init` in repro generator

kylegach commented:
_A note on workflow:_

As of this moment, here is the proposed workflow for this work:

1. Continue addressing feedback in this PR
1. Get all non-repro/sandbox CI checks green
1. Run the [Generate and push repros to the next branch workflow](https://github.com/storybookjs/storybook/actions/workflows/generate-repros-next.yml) manually, on this branch
    - This ensures an updated version of the repros are available at https://github.com/storybookjs/repro-templates-temp
1. Once successful, merge this PR
1. Re-run checks on https://github.com/storybookjs/storybook/pull/19478
1. Address any bugs/feedback
1. Merge that PR once green

lifeiscontent commented on code/frameworks/nextjs/package.json:
```diff
@@ -0,0 +1,129 @@
+  "publishConfig": {
+    "access": "public"
+  },
+  "bundler": {
+    "entries": [
+      "./src/index.ts",
+      "./src/preset.ts",
+      "./src/preview.tsx",
+      "./src/next-image-loader-stub.ts",
+      "./src/resolved-router-context.ts"
+    ],
+    "platform": "node"
```
what does this stuff mean? is this exclusive to storybook or webpack?

kylegach replied:
It's used by tsup, which is what Storybook uses to bundle (and pre-bundle) various packages.

yannbf committed:
fix helpers test

yannbf committed:
temporarily fix tests

yannbf committed:
fix lint issues in nextjs framework

yannbf committed:
show logs in repros generation action

yannbf committed:
switch order of repro templates

yannbf committed:
Merge branch 'next' into nextjs-framework-support

yannbf committed:
fix import

yannbf committed:
Merge branch 'next' into nextjs-framework-support

yannbf committed:
remove .git folder when generating repros

yannbf committed:
provide babel preset typescript out of the box

yannbf committed:
Merge branch 'next' into nextjs-framework-support

yannbf committed:
update package version

yannbf committed:
fix prep command to get listed by nx

yannbf committed:
skip tsconfig-paths if projects contain no baseUrl

yannbf committed:
set correct expected framework in repro-templates for cra

yannbf committed:
include framework specific stories in sandbox

yannbf committed:
add nextjs specific stories in sandbox

yannbf committed:
fix vue2 repro framework

yannbf committed:
fix lint issues

yannbf committed:
fix lint issue

yannbf committed:
Merge branch 'next' into nextjs-framework-support

yannbf committed:
remove debug flag

yannbf committed:
support ESM nextjs config files

yannbf committed:
add comment in nextjs css webpack config for future reference

yannbf committed:
skip lib check in nextjs framework package

yannbf committed:
fix nextConfigFile logic

yannbf committed:
only inject babel typescript preset in typescript projects

shilman reviewed (approved):
🎉 🎉 🎉

bertho-zero commented:
It looks like there is no support for `next/future/image` (https://github.com/RyanClementsHax/storybook-addon-next/issues/99)

shilman commented:
@kylegach @yannbf let's go through the open issues on the addon & see what we want to address in the framework for 7.0?

RyanClementsHax commented:
> @kylegach @yannbf let's go through the open issues on the addon & see what we want to address in the framework for 7.0?

I'd recommend doing this and also going through the closed issues. Some of the issues opened were valid but were automatically closed because no activity was recorded on them for some time.

kylegach commented:
> It looks like there is no support for `next/future/image` ([RyanClementsHax/storybook-addon-next#99](https://github.com/RyanClementsHax/storybook-addon-next/issues/99))

@bertho-zero — It was decided to ship what we have ready (and what `storybook-addon-next` currently supports) and then add further features like that. We also need to coordinate with the Next.js folks about the best way to read the Next config. We definitely haven't forgotten about it!
